src/Servers/IIS/IIS/test/Common.FunctionalTests/Inprocess/ResponseTrailersTests.cs

```diff
@@ -0,0 +1,186 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Globalization;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Server.IIS.FunctionalTests.Utilities;
+using Microsoft.AspNetCore.Server.IntegrationTesting.Common;
+using Microsoft.AspNetCore.Server.IntegrationTesting.IIS;
+using Microsoft.AspNetCore.Testing;
+using Microsoft.Net.Http.Headers;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Server.IIS.FunctionalTests
+{
+    [Collection(PublishedSitesCollection.Name)]
+    public class ResponseTrailersTests : IISFunctionalTestBase
```

Comment:
Please add a test that combines trailers with CompleteAsync.

Check that CompleteAsync immediately sends trailers to the client and ends the response while the request delegate is still running.

Reply:
Those are the tests we had to remove, CompleteAsync doesn't work in IIS's request lifetime model. Is that going to be a problem? Maybe for the streaming error case?

Reply:
CompleteAsync is used with the deadline timeout. When the timeout is reached the timer calls CompleteAsync to send trailing HEADERS frame, then sends RST_STREAM frame with either IHttpResetFeature/HttpContext.Abort.

What will happen in IIS?

Reply:
The reset will probably happen first and prevent the trailers from being sent.

Reply:
FYI Justin is implementing reset now so we should be able to verify that.

Reply:
Will the reset/abort happen immediately or will it wait for the request delegate to complete?

Would properly supporting CompleteAsync require changes in IIS? How big a change would it be?

Does HttpSys properly support CompleteAsync?

Reply:
HttpSys does properly support CompleteAsync.

Justin has some ideas for how to implement CompleteAsync in IIS. There's a related issue over at https://github.com/dotnet/aspnetcore/issues/17268. I think it can be done without changes to IIS, but it would require a lot of hardening to AspNetCore to avoid touching any native IIS structures after CompletAsync.

Reply:
Ok! To give you an idea of what the impact on gRPC would be:

If CompleteAsync isn't supported correctly, and the server doesn't send trailers to the client, there is a chance that gRPC calls that exceed the deadline timeout will fail with the wrong status code. I say chance because there is always a race between the client and server to report the exceeded deadline. If the client wins then the right status will be reported, if the server wins then the wrong status will be reported.

This wouldn't be disastrous to gRPC support on IIS, but it would be a known issue that would have to be called out.